src/ray/gcs/gcs_server/gcs_resource_scheduler.cc

```diff
@@ -0,0 +1,301 @@
+  result.resize(required_resources_list.size());
+  absl::flat_hash_set<NodeID> candidate_nodes_copy(candidate_nodes);
+  std::list<std::pair<int, ResourceSet>> required_resources_list_copy;
+  int index = 0;
+  for (const auto &iter : required_resources_list) {
+    required_resources_list_copy.emplace_back(index++, iter);
+  }
+
+  while (!required_resources_list_copy.empty()) {
+    // Score and sort nodes.
+    const auto &required_resources_index = required_resources_list_copy.front().first;
+    const auto &required_resources = required_resources_list_copy.front().second;
+    const auto &node_scores = ScoreNodes(required_resources, candidate_nodes_copy);
+    if (node_scores.empty() || node_scores.front().second < 0) {
+      // There is no node to meet the scheduling requirements.
+      break;
+    }
+
+    const auto &highest_score_node_id = node_scores.front().first;
+    RAY_CHECK(gcs_resource_manager_.AcquireResources(highest_score_node_id,
+                                                     required_resources));
+    result[required_resources_index] = highest_score_node_id;
+    required_resources_list_copy.pop_front();
+
+    // We try to schedule more resources on one node.
+    for (auto iter = required_resources_list_copy.begin();
+         iter != required_resources_list_copy.end();) {
+      if (gcs_resource_manager_.AcquireResources(highest_score_node_id, iter->second)) {
+        result[iter->first] = highest_score_node_id;
+        required_resources_list_copy.erase(iter++);
+      } else {
+        ++iter;
+      }
+    }
+    candidate_nodes_copy.erase(highest_score_node_id);
+  }
+
+  // Releasing the resources temporarily deducted from `gcs_resource_manager_`.
+  ReleaseTemporarilyDeductedResources(required_resources_list, result);
+
+  if (!required_resources_list_copy.empty()) {
+    // Unable to meet the resources required for scheduling, scheduling failed.
+    result.clear();
+  }
+  return result;
+}
+
+std::list<NodeScore> GcsResourceScheduler::ScoreNodes(
+    const ResourceSet &required_resources,
+    const absl::flat_hash_set<NodeID> &candidate_nodes) {
+  std::list<NodeScore> node_scores;
+  const auto &cluster_resources = gcs_resource_manager_.GetClusterResources();
+
+  // Score the nodes.
+  for (const auto &node_id : candidate_nodes) {
+    const auto &iter = cluster_resources.find(node_id);
+    RAY_CHECK(iter != cluster_resources.end());
+    double node_grade = node_scorer_->Grade(required_resources, iter->second);
+    node_scores.emplace_back(node_id, node_grade);
+  }
+
+  // Sort node scores, the large score is in the front.
+  node_scores.sort([](const NodeScore &left, const NodeScore &right) {
+    return right.second < left.second;
+  });
+  return node_scores;
+}
+
+void GcsResourceScheduler::ReleaseTemporarilyDeductedResources(
+    const std::vector<ResourceSet> &required_resources_list,
+    const std::vector<NodeID> &nodes) {
+  for (int index = 0; index < (int)nodes.size(); ++index) {
```

Comment:
```suggestion
  for (int index = 0; index < (int)nodes.size(); index++) {
```
For consistency (I believe there's no reason to use ++index here. If there's a reason lmk.)